packages/editor/src/components/post-publish-panel/postpublish.js

```diff
@@ -41,142 +41,131 @@ const getFuturePostUrl = ( post ) => {
 	return post.permalink_template;
 };
 
-function CopyButton( { text, onCopy, children } ) {
-	const ref = useCopyToClipboard( text, onCopy );
-	return (
-		<Button __next40pxDefaultSize variant="secondary" ref={ ref }>
-			{ children }
-		</Button>
-	);
-}
-
-class PostPublishPanelPostpublish extends Component {
-	constructor() {
-		super( ...arguments );
-		this.state = {
-			showCopyConfirmation: false,
+export default function PostPublishPanelPostpublish( {
+	focusOnMount,
+	children,
+} ) {
+	const { post, postType, isScheduled } = useSelect( ( select ) => {
+		const {
+			getEditedPostAttribute,
+			getCurrentPost,
+			isCurrentPostScheduled,
+		} = select( editorStore );
+		const { getPostType } = select( coreStore );
+
+		return {
+			post: getCurrentPost(),
+			postType: getPostType( getEditedPostAttribute( 'type' ) ),
+			isScheduled: isCurrentPostScheduled(),
 		};
-		this.onCopy = this.onCopy.bind( this );
-		this.onSelectInput = this.onSelectInput.bind( this );
-		this.postLink = createRef();
-	}
-
-	componentDidMount() {
-		if ( this.props.focusOnMount ) {
-			this.postLink.current.focus();
+	}, [] );
+	const [ showCopyConfirmation, setShowCopyConfirmation ] = useState( false );
+	const timeoutIdRef = useRef();
+
+	useEffect( () => {
+		return () => {
+			if ( timeoutIdRef.current ) {
+				clearTimeout( timeoutIdRef.current );
+			}
+		};
+	}, [] );
+
+	const postLabel = postType?.labels?.singular_name;
+	const viewPostLabel = postType?.labels?.view_item;
+	const addNewPostLabel = postType?.labels?.add_new_item;
+	const link =
+		post.status === 'future' ? getFuturePostUrl( post ) : post.link;
+	const addLink = addQueryArgs( 'post-new.php', {
+		post_type: post.type,
+	} );
+
+	const copyRef = useCopyToClipboard( link, () => {
+		setShowCopyConfirmation( true );
+		if ( timeoutIdRef.current ) {
+			clearTimeout( timeoutIdRef.current );
 		}
-	}
-
-	componentWillUnmount() {
-		clearTimeout( this.dismissCopyConfirmation );
-	}
-
-	onCopy() {
-		this.setState( {
-			showCopyConfirmation: true,
-		} );
-
-		clearTimeout( this.dismissCopyConfirmation );
-		this.dismissCopyConfirmation = setTimeout( () => {
-			this.setState( {
-				showCopyConfirmation: false,
-			} );
+		timeoutIdRef.current = setTimeout( () => {
+			setShowCopyConfirmation( false );
 		}, 4000 );
-	}
+	} );
+	const postLinkRef = useCallback(
+		( node ) => {
+			if ( focusOnMount && node ) {
+				node.focus();
+			}
+		},
+		[ focusOnMount ]
+	);
```

Comment:
It's better to avoid ref callback if possible. The usual pattern is to have `postLinkRef` as a regular ref (`useRef`) and do the focusing in `useEffect`.

Reply:
What's the downside here? Seems like a good pattern for simple "focus on mount" cases.

Ref: https://tkdodo.eu/blog/avoiding-use-effect-with-callback-refs

Reply:
There is no downside, there is a popular article about it (TIL), and also Gutenberg's `useRefEffect` helper uses this pattern all the time. It's just that IMO the idiomatic way to focus on mount is ref+effect 🙂

Reply:
Do you mind if we keep this pattern here? I'm happy to follow up if it becomes a problem at some point.

Reply:
No problem, keep it if you like it 🙂